src/main/java/core/basesyntax/Car.java

```diff
@@ -1,13 +1,88 @@
 package core.basesyntax;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 
 /**
  * Make this class immutable. See requirements in the README file
  */
-public class Car {
-    private int year;
-    private String color;
-    private List<Wheel> wheels;
-    private Engine engine;
+public final class Car {
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = deepWheelCopy(wheels);
+        this.engine = engineClone(engine);
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(year, color, wheels, engine);
+    }
+
+    public Car addWheel(Wheel wheel) {
+        List<Wheel> newWheels = getWheels();
+        newWheels.add(wheel);
+        return new Car(year,color, newWheels, engine);
+    }
+
+    public Car changeColor(String color) {
+        return new Car(year, color, wheels, engine);
+    }
+
+    private List<Wheel> deepWheelCopy(List<Wheel> wheels) {
+        List<Wheel> copiedWheels = new ArrayList<Wheel>();
+        for (Wheel wheel : wheels) {
+            copiedWheels.add(wheel.clone());
+        }
+        return copiedWheels;
+    }
+
+    private Engine engineClone(Engine engine) {
+        if (engine != null) {
+            return engine.clone();
+        } else {
+            return null;
+        }
```

Comment:
```suggestion
        } 
        return null;
```
or 
return engine == null ? null : engine.clone();